eq3bt/connection.py

```diff
@@ -15,46 +17,83 @@
 class BTLEConnection(btle.DefaultDelegate):
     """Representation of a BTLE Connection."""
 
-    def __init__(self, mac):
+    def __init__(self, mac, keep_connected=False):
         """Initialize the connection."""
         btle.DefaultDelegate.__init__(self)
 
+        self._ifaces = self.get_hci_ifaces()
+        self._iface_idx = 0
+
         self._conn = None
         self._mac = mac
         self._callbacks = {}
+        self._keep_connected = keep_connected
+    
+    def next_iface(self):
+        self._nr_conn_errors += 1
+        self._iface_idx = (self._iface_idx + 1) % len(self._ifaces)
+        if self._nr_conn_errors >= len(self._ifaces)*2:
+            return False
+        return True
+
+    def get_hci_ifaces(self):
+        iface_list = []
+        bus = dbus.SystemBus()
+        manager = dbus.Interface(bus.get_object("org.bluez", "/"), "org.freedesktop.DBus.ObjectManager")
+        objects = manager.GetManagedObjects()
+        for path, interfaces in objects.items():
+            adapter = interfaces.get("org.bluez.Adapter1")
+            if adapter is None:
+                continue
+            iface_list.append(re.search(r'\d+$', path)[0])
+        return iface_list
 
     def __enter__(self):
         """
         Context manager __enter__ for connecting the device
         :rtype: btle.Peripheral
         :return:
         """
-        self._conn = btle.Peripheral()
-        self._conn.withDelegate(self)
-        _LOGGER.debug("Trying to connect to %s", self._mac)
         try:
-            self._conn.connect(self._mac)
-        except btle.BTLEException as ex:
-            _LOGGER.debug("Unable to connect to the device %s, retrying: %s", self._mac, ex)
-            try:
-                self._conn.connect(self._mac)
-            except Exception as ex2:
-                _LOGGER.debug("Second connection try to %s failed: %s", self._mac, ex2)
-                raise
+            conn_state = self._conn.getState()
+        except:
+            self._conn = None
```

Comment:
This will lose the reference to the potentially already existing `_conn` object, is there something else that should be done in cases where `getState` errors on already existing connection?

Anycase, please don't use bare excepts but catch only exceptions you are expecting.

Reply:
Ok, I can change that to catch only the errors which indicate that the connection was already lost and therefore a disconnect() call isn't needed.